### DISCUSSION TITLE
does taskinstance.py execute __del__() function for instantiated object during Operator execution.

### DISCUSSION THREAD

**Discussion by kiwy42** [apache/airflow#18145] [Q&A]
Hello,   
I'm trying to use properly a [hdfs_client](https://github.com/gupta-paras/pyhdfs_client).
This is a py4j implementation of hdfs command which mean it opens file local socket to java gateway and so on .
I've build a small module (singleton more or less) with a `__new__(cls)` a `__del__()` func ,   
I've put in a module I import in dag where I need to interact with HDFS.  
If you ever used py4j you know it creates tmp files, socket and lots of stuff and you want to clean it up when done.  
However I think the method `__del__()` is never called as I see my `/tmp` flooding with temp files.

Can someone with a better comprehension of Airflow task system tel me how I could insure my singleton is properly terminated ? 
Because it also seems that the Singleton is shutdown between task according to my logs.

This might be a bit confusing I would be glad if someone could help me with that, I'm pretty sure I'm not the only one to use external tools that generates temps file and need to clean them before my task is done. If I need to do it manually I will but it's strange as there's a proper __del__() method define in my Class.

**Comment by potiuk** [comment]
First of all a piece of advice: never, ever use `__del__` except in very low-level functions. You have no control on when `__del__` is called  and there are some cases that if you not implement `__del__` properly,  it might severly impact performance of reference counting and cyclic garbage collector.

Secondly - see all the warnings and notes in https://docs.python.org/3/reference/datamodel.html#object.__del__  - one of the points you will see there that `__del__`` method is NOT guaranteed to be called when the interpreter terminates.  This is likely what happens - because ...

Airflow parses each DAG file every 30 seconds or so (by default) - and it instantiates the DAG in separate process and sandbox every time it does it. See https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code  - you should only do "serious" stuff inside your operator's "execute" method or in JINJA macros.

**Comment by potiuk** [comment]
Instead you should instantiate your HDFS client in "execute" method and implement `try ... finally` to delete whatever needs to be deleted.

**Comment by kiwy42** [comment]
First of all thank you so much for the support, you already helped me with my first PR and now that.
That's kind of stuff is amazing.

My module looks like this (over simplified): 

```

import pyhdfs_client

Class HdfsFunction(object):
	instance = None
	client_instance = None
	
	def __new__(cls):
		cls.instance = super.object.__new__()
			
	def __del__(cls):
		cls.client_instance.stop()
		cls.client_instance = None

    def getglient(cls):
		if cls.instance.client_instance == None:
			cls.instance.client_instance = pyhdfs_client()
		return cls.instance.client_instance = pyhdfs_client()
	
	def list_dir(cls): 
		ret = cls.getclient().listdir()
		return ret
```

This is over simplified of course, however for sake of readability I will not post all the code.  
The idea is that I import this class inside my dag when task is running, the client is instanciate and run as long as the task is running (like some kind of a java singleton).  
in my dag I do:

```
import HdfsFunction

def a_python_operator_task(): 
	mydirlist = HdfsFunction.list_dir()
	mydirlist2 = HdfsFunction.list_dir()
	[...]
	# and so on up to the end of my task code
```

In such case I'm not sure to understand where should I put the try catch finally statement.  
If I follow you I should do such check in my task instead of my module ?   
My idea here is that I want to leave open my hdfs client as long as possible and reuse it as loading py4j are very slow to init and if possible remove all possibility for other developper to misuse the client inside there own dag.  

I understand that those are not necessarilly Airflow specific question and more Python 101, but coming from java and scala, python is sometime hard to follow (specially in the context of a complex workflow framework).  

In any case thank you so very much for the help.
